### PR TITLE
ec2_vol_facts: moved to boto3

### PR DESCRIPTION
**Pull request by hryamzik**
##### SUMMARY

Moved to boto3 to support role-based profiles.

##### ISSUE TYPE

 - Bugfix Pull Request

##### COMPONENT NAME

ec2_vol_facts

##### ANSIBLE VERSION

```
2.6.1
```

### CODE REVIEW

**Comment by hryamzik** [comment]
Love these formatting requirements, 2 imports instead of a single multiline one.

```python
from ansible.module_utils.ec2 import connect_to_aws, ec2_argument_spec, get_aws_connection_info, boto3_conn, HAS_BOTO3, boto3_tag_list_to_ansible_dict
from ansible.module_utils.ec2 import ansible_dict_to_boto3_filter_list, camel_dict_to_snake_dict
```

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vol_facts.py]: boto3_conn is calling module.fail_json for exceptions and the lack of region, so you can remove the exception handling and unnest this and remove the ProfileNotFound import

**Comment by hryamzik** [comment]
@s-hertel is there anything preventing this from being merged?

**Comment by hryamzik** [comment]
cc @mattclay

Is there any way to avoid stuck PRs like this?

**Comment by mattclay** [comment]
@hryamzik Since this is a community module, the best thing to do is try pinging the maintainer of the module for a review. If the maintainer is unavailable, you may be able to find others in the community to review and test. Since this is an AWS module you may want to check in the #ansible-aws IRC channel on Freenode.

**Comment by hryamzik** [comment]
Hey @wimnat, can you take a look at this PR, please?

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vol_facts.py:63]: This module should handle BotoCoreError and ClientError. ProfileNotFound should be handled by boto3_conn and doesn't need to be imported by the module.

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vol_facts.py:68]: You could update this module to AnsibleAWSModule if you wanted. That would simplify correcting the exception handling and eliminate the need to use/import all of:
`connect_to_aws`, `ec2_argument_spec`, `get_aws_connection_info`, `boto3_conn`, and `HAS_BOTO3`

Doing this would end up looking something like:
```
from ansible.module_utils.aws.core import AnsibleAWSModule
...
module = AnsibleAWSModule(argument_spec=argument_spec)
connection = module.client('ec2')
```

And your exception handling would look like:
```
try:
    connection.describe_volumes(Filters=ansible_dict_to_boto3_filter_list(sanitized_filters))
except (BotoCoreError, ClientError) as e:
    module.fail_json_aws(e, msg="Unable to describe volumes with filters {0}".format(sanitized_filters))
```

Otherwise you end up having a mess handling the different cases of what attributes exist for what exception because they don't inherit the same properties (such as, you should include e.message and e.response if they exist, but e.message does not exist on Python 3 and BotoCoreErrors never have a response attribute).

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vol_facts.py]: You've imported boto3_tag_list_to_ansible_dict so you could use that.
line 99: `'tags': boto3_tag_list_to_ansible_dict(volume['tags'])`

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vol_facts.py]: Do you need `camel_dict_to_snake_dict(volume, ignore_list=['Tags'])`? If 'Tags' is a key in volumes it will snake_dict the tags too.

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vol_facts.py]: It would be nice to have integration tests to make sure this module stays backward compatible. It's not a blocker, but there *should* be tests for features even for preview modules. https://github.com/ansible/community/blob/master/group-aws/integration.md#policy

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vol_facts.py]: This should handle pagination https://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Paginator.DescribeVolumes

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vol_facts.py]: e.message does not exist on Python 3. e.response does exist for ClientErrors and should be included if so.

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (hryamzik) [reply lib/ansible/modules/cloud/amazon/ec2_vol_facts.py]: I don't see any pagination support in the original code, is it something different in `boto3` approach?

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (hryamzik) [reply lib/ansible/modules/cloud/amazon/ec2_vol_facts.py]: I'm working on an integration test but it's hard to compare everything so let me post results here:

current version:

```
    "volumes": [
        {
            "attachment_set": {
                "attach_time": "2018-02-12T12:43:22.000Z",
                "device": "/dev/sda1",
                "instance_id": "i-88888888888888885",
                "status": "attached"
            },
            "create_time": "2018-02-12T12:43:22.461Z",
            "encrypted": false,
            "id": "vol-88888888888888888",
            "iops": 100,
            "region": "us-west-2",
            "size": 8,
            "snapshot_id": "snap-88888888888888887",
            "status": "in-use",
            "tags": {
                "Name": "test",
            },
            "type": "gp2",
            "zone": "us-west-2c"
        }
]
```

new version:

```
    "volumes": [
        {
            "attachment_set": {
                "attach_time": "2018-02-12T12:43:22+00:00", 
                "delete_on_termination": true, 
                "device": "/dev/sda1", 
                "instance_id": "i-88888888888888885", 
                "status": "attached"
            }, 
            "create_time": "2018-02-12T12:43:22.461000+00:00", 
            "encrypted": false, 
            "id": "vol-88888888888888888", 
            "iops": 100, 
            "region": "us-west-2", 
            "size": 8, 
            "snapshot_id": "snap-88888888888888887", 
            "status": "in-use", 
            "tags": {
                "Name": "test", 
            }, 
            "type": "gp2", 
            "zone": "us-west-2c"
        },
]
```

So as you can see date format is a bit different, is it something that should be checked/fixed?

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (hryamzik) [reply lib/ansible/modules/cloud/amazon/ec2_vol_facts.py]: Good point, I didn't manage find any tags issues but going to give this change a try.

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (hryamzik) [reply lib/ansible/modules/cloud/amazon/ec2_vol_facts.py:68]: I didn't manage to get value for region in this case and it has to be presented in module output.

**Comment by hryamzik** [comment]
>ClientError: An error occurred (UnauthorizedOperation) when calling the RunInstances operation: You are not authorized to perform this operation

This is wired as I've copied setup from other module test. Any ideas how to fix this?

File: lib/ansible/modules/cloud/amazon/ec2_vol_facts.py
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/ec2_vol_facts.py:68]: Okay, no problem either way. You could still get region via get_aws_connection_info, which is how AnsibleAWSModule gets it too.

**Review by ryansb — APPROVED** [review]
LGTM

File: test/integration/targets/ec2_vol_facts/tasks/main.yml
Comment (ryansb) [inline test/integration/targets/ec2_vol_facts/tasks/main.yml]: This (and any other resources) need to be using `{{ resource_prefix }}` on resource names, it's part of the requirements for the CI IAM policy.

File: test/integration/targets/ec2_vol_facts/tasks/main.yml
Comment (ryansb) [inline test/integration/targets/ec2_vol_facts/tasks/main.yml]: Same here - make sure to use `{{ resource_prefix }}`

**Review by ryansb — CHANGES_REQUESTED** [review]

File: test/integration/targets/ec2_vol_facts/defaults/main.yml
Comment (willthames) [inline test/integration/targets/ec2_vol_facts/defaults/main.yml]: I prefer not to include effectively empty defaults and vars files

File: test/integration/targets/ec2_vol_facts/tasks/main.yml
Comment (willthames) [inline test/integration/targets/ec2_vol_facts/tasks/main.yml:77]: This test fails - this is the only reference to this tag.

**Review by willthames — CHANGES_REQUESTED** [review]

File: test/integration/targets/ec2_vol_facts/aliases
Comment (mattclay) [inline test/integration/targets/ec2_vol_facts/aliases:2]: It looks like testing this PR is blocked on CI updates that @s-hertel is working on.

We should wait for those updates rather than disabling the tests.

**Comment by s-hertel** [comment]
+1 to merge from me, and I will update CI for these when I have time. The tests look good and pass locally for me with Python 2 and 3. And it's pretty low risk since it's a preview community facts module.

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
Merged, thanks @hryamzik

**Comment by willthames** [comment]
Thanks to @s-hertel for the secondary testing too

**Comment by hryamzik** [comment]
Thanks guys!

**Pull request merged**
